### ISSUE TITLE
Unclear what the type coercion rules are

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap--Build-Your-Own-Blocks#2112]
I would have expected all of the following to have the same truth value.

![image](https://user-images.githubusercontent.com/2138696/40926490-3d54b9c6-6814-11e8-8255-0b43ea195ac2.png)

Sometimes it always converts text to numbers even if there are spaces on either side of the text. But it seems = sometimes corerces while + always does. Is there a general rule?

**Comment by jguille2** [comment]
Yes, sometimes this type conversion may confuse. We have a non-typed system, but really we have strings and numbers!

Usually, conversion rules are quite clear, but not always... but we can use "math operators" and "string operators" to force some changes be sure of the result.

In fact, the "comparison operators" seems like a non-stricted comparison... but this is not the truth: you can see [snapEquals function](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/blob/master/threads.js#L74) and we have an strict comparison after some changes... These changes make a `case-insensitive` comparison and also a `string->number conversion`... but **this conversion does not allow whitespaces_**(I think this is the great confusion here).

Then...

In the first two cases, we have only the _comparison operator_, and it is running as I've commented. The block try to convert the string into a number. And yes, `"12"->12` (1st case). And it would also go well with other cases like `"+12"->12`, or `"12.0"->12`. But, as I said, _whitespaces_ are not allowed, and then `"   12"` is not converted to `12`.

But int the third case, we have a _math operator_ and then, it try to convert strings to numbers in the normal way (here _whitespaces_ are allowed). So the conversion is not into the _comparison_, is into the _sum_. Then `"     12     " + 0`  is converted to `12 + 0`. And after this, the _comparison block_ has to do nothing else: `12 ===  12` is true.

**Comment by cycomachead** [comment]
The type coercion inherits from Javascript. Plus it includes some stuff Scratch invented. IMO, this is one of the worst aspects of Snap! as a programming language. 

In this case, the behavior is definitely confusing. I didn't expect the 3rd to work initially, but it did, and it turns out JS doesn't care about leading or trailing whitespace around numbers.

The second seems to because by Snap! checking for whitespace when dealing with it's numeric conversion in `snapEquals`. ([line 99](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/blob/master/threads.js#L99)). To be honest, I don't quite understand why. I get protecting against special values, but I don't see what the extra complexity is preventing.
